src/walletConnect/request.ts

```diff
@@ -4,35 +4,84 @@ import { ContractKit } from '@celo/contractkit'
+import networkConfig from 'src/web3/networkConfig'
 import { getWalletAddress, unlockAccount } from 'src/web3/saga'
 import { applyChainIdWorkaround, buildTxo } from 'src/web3/utils'
 import { call } from 'typed-redux-saga'
-import Web3 from 'web3'
+import {
+  Address,
+  GetTransactionCountParameters,
+  SignMessageParameters,
+  formatTransaction,
+} from 'viem'
+import { getTransactionCount } from 'viem/actions'
 
 const TAG = 'WalletConnect/handle-request'
 
-export interface WalletResponseError {
-  isError: true
-  error: string
-}
-export interface WalletResponseSuccess {
-  isError: false
-  result: string
-}
-
 export function* handleRequest({ method, params }: { method: string; params: any[] }) {
-  const account: string = yield* call(getWalletAddress)
-  const wallet: UnlockableWallet = yield* call(getWallet)
+  const wallet: ViemWallet = yield* call(getViemWallet, networkConfig.viemChain.celo)
+  const useViem = yield* call(
+    getFeatureGate,
+    StatsigFeatureGates.USE_VIEM_FOR_WALLETCONNECT_TRANSACTIONS
+  )
+
+  const account = yield* call(getWalletAddress)
+  const legacyWallet: UnlockableWallet = yield* call(getWallet)
   yield* call(unlockAccount, account)
   // Call Sentry performance monitoring after entering pin if required
   SentryTransactionHub.startTransaction(SentryTransaction.wallet_connect_transaction)
 
   switch (method) {
     case SupportedActions.eth_signTransaction: {
+      // TODO: keep only Viem branch after feeCurrecny estimation is ready
+      if (useViem) {
+        const rawTx: any = { ...params[0] }
+
+        let tx: any
+        // Provide an escape hatch for dapp developers who don't want any normalization
+        if (rawTx.__skip_normalization) {
+          delete rawTx.__skip_normalization
+          tx = rawTx
+        } else {
+          tx = yield* call(normalizeTransaction, wallet, rawTx)
+        }
+
+        // Convert hex values to numeric ones for Viem
+        // TODO: remove once Viem allows hex values as quanitites
+        const formattedTx: any = yield* call(formatTransaction, tx)
+
+        // TODO: inject fee currency for Celo when it's ready
+
+        const txCountParams: GetTransactionCountParameters = {
+          address: account as Address,
+          blockTag: 'pending',
+        }
+        const nonce = yield* call(getTransactionCount, wallet, txCountParams)
+        const { maxFeePerGas, maxPriorityFeePerGas } = yield* call(
+          estimateFeesPerGas,
+          wallet,
+          formattedTx.feeCurrency
+        )
+        const txRequest = {
+          ...formattedTx,
+          ...(formattedTx.nonce === undefined ? { nonce } : {}),
+          ...(formattedTx.maxFeePerGas === undefined ? { maxFeePerGas } : {}),
+          ...(formattedTx.maxPriorityFeePerGas === undefined ? { maxPriorityFeePerGas } : {}),
+        }
+
+        return (yield* call([wallet, 'signTransaction'], txRequest)) as string
```

Comment:
the logic in this function looks good to me, although i wonder if there's a way to avoid all the `any`'s? it's an annoying question but i couldn't help but think that anyone touching this has to be really confident in the changes being made, because there's not much that typescript is going to do to help you here 😅 maybe it's unavoidable or would add too much complexity, i didn't think this suggestion through too deeply but just wanted to make a note 😅

Reply:
I wish I knew that way 😭

Unfortunately I can't figure out a proper type orchestration for Viem+Celo. It's too smart for me.

Reply:
The problem seems to be global: https://valora-app.slack.com/archives/C029Z1QMD7B/p1698657194032799

My proposal it to keep `any`s for now and refactor once we find the solution for proper typing. Probably along with [RET-905](https://linear.app/valora/issue/RET-905/[walletconnect]-add-feecurrency-injection-back)